File: backend/agent_service/surface_validator.py
"""Closed structural and semantic validation for SurfaceSpec v0."""

import json
import re
from pathlib import Path


ROOT = Path(__file__).resolve().parents[2]
NODE_ID = re.compile(r"^[A-Za-z][A-Za-z0-9_-]{0,63}$")
LAYOUT_TYPES = {"Row", "Column"}
LAYOUT_KEYS = {"id", "type", "children", "gap", "align", "justify", "weight"}
BUSINESS_KEYS = {"id", "type", "weight"}
ROOT_KEYS = {"version", "surfaceId", "root", "nodes"}
GAPS = {"none", "small", "medium", "large"}
ALIGNS = {"start", "center", "end", "stretch"}
JUSTIFIES = {"start", "center", "end", "spaceBetween", "spaceAround", "spaceEvenly"}


def _catalog_types(catalog_path=None):
    path = catalog_path or ROOT / "shared" / "catalog" / "component-catalog.json"
    catalog = json.loads(Path(path).read_text(encoding="utf-8"))
    return {
        item["type"]: bool(item["multiple"])
        for item in catalog["components"]
    }


def validate_surface_spec(document, max_nodes=32, max_depth=8, catalog_path=None):
    """Return a stable list of diagnostics; an empty list means valid."""
    errors = []
    if not isinstance(document, dict):
        return ["document must be an object"]
    if set(document) != ROOT_KEYS:
        errors.append("document fields must be exactly version, surfaceId, root, nodes")
    if document.get("version") != "0.1":
        errors.append("version must be 0.1")
    if document.get("surfaceId") != "main":
        errors.append("surfaceId must be main")

    root_id = document.get("root")
    if not isinstance(root_id, str) or not NODE_ID.match(root_id):
        errors.append("root must be a valid node ID")
    nodes = document.get("nodes")
    if not isinstance(nodes, list):
        return errors + ["nodes must be an array"]
    if len(nodes) > max_nodes:
        errors.append("node limit exceeded")

    catalog = _catalog_types(catalog_path)
    by_id = {}
    type_counts = {}
    children_by_id = {}
    for index, node in enumerate(nodes):
        label = "nodes[{}]".format(index)
        if not isinstance(node, dict):
            errors.append(label + " must be an object")
            continue
        node_id = node.get("id")
        node_type = node.get("type")
        if not isinstance(node_id, str) or not NODE_ID.match(node_id):
            errors.append(label + " has an invalid id")
            continue
        if node_id in by_id:
            errors.append("duplicate node id: " + node_id)
        by_id[node_id] = node
        if node_type in LAYOUT_TYPES:
            if not {"id", "type", "children"}.issubset(node):
                errors.append(label + " is missing layout fields")
            if not set(node).issubset(LAYOUT_KEYS):
                errors.append(label + " contains forbidden layout fields")
            children = node.get("children")
            if not isinstance(children, list) or any(not isinstance(item, str) for item in children):
                errors.append(label + ".children must be an array of IDs")
                children = []
            if len(children) > max_nodes:
                errors.append(label + ".children exceeds the node limit")
            children_by_id[node_id] = children
            if "gap" in node and node["gap"] not in GAPS:
                errors.append(label + " has an invalid gap")
            if "align" in node and node["align"] not in ALIGNS:
                errors.append(label + " has an invalid align")
            if "justify" in node and node["justify"] not in JUSTIFIES:
                errors.append(label + " has an invalid justify")
        elif node_type in catalog:
            if not {"id", "type"}.issubset(node):
                errors.append(label + " is missing business fields")
            if not set(node).issubset(BUSINESS_KEYS):
                errors.append(label + " contains forbidden business fields")
            type_counts[node_type] = type_counts.get(node_type, 0) + 1
            children_by_id[node_id] = []
        else:
            errors.append(label + " has an unknown or forbidden type")
            children_by_id[node_id] = []
        if "weight" in node:
            weight = node["weight"]
            if isinstance(weight, bool) or not isinstance(weight, int) or weight < 0 or weight > 10:
                errors.append(label + " has an invalid weight")

    if isinstance(root_id, str) and root_id not in by_id:
        errors.append("root does not reference a declared node")
    elif root_id in by_id and "weight" in by_id[root_id]:
        errors.append("root must not declare weight")

    parent_count = {node_id: 0 for node_id in by_id}
    for parent_id, children in children_by_id.items():
        parent = by_id.get(parent_id, {})
        has_positive_weight = False
        for child_id in children:
            if child_id not in by_id:
                errors.append(parent_id + " references unknown child " + child_id)
                continue
            parent_count[child_id] += 1
            if by_id[child_id].get("weight", 0) > 0:
                has_positive_weight = True
        if has_positive_weight and parent.get("justify", "start") != "start":
            errors.append(parent_id + " combines positive child weight with non-start justify")

    for node_id, count in parent_count.items():
        expected = 0 if node_id == root_id else 1
        if count != expected:
            errors.append(node_id + " has {} parents; expected {}".format(count, expected))

    visited = set()
    active = set()

    def visit(node_id, depth):
        if node_id in active:
            errors.append("cycle detected at " + node_id)
            return
        if node_id in visited or node_id not in by_id:
            return
        if depth > max_depth:
            errors.append("depth limit exceeded at " + node_id)
        active.add(node_id)
        for child_id in children_by_id.get(node_id, []):
            visit(child_id, depth + 1)
        active.remove(node_id)
        visited.add(node_id)

    if root_id in by_id:
        visit(root_id, 1)
    unreachable = set(by_id) - visited
    if unreachable:
        errors.append("unreachable nodes: " + ", ".join(sorted(unreachable)))

    for component_type, multiple in catalog.items():
        if not multiple and type_counts.get(component_type, 0) > 1:
            errors.append(component_type + " does not allow multiple instances")
    return errors


def load_and_validate_surface_spec(path, max_nodes=32, max_depth=8):
    try:
        document = json.loads(Path(path).read_text(encoding="utf-8"))
    except (OSError, ValueError) as error:
        return ["invalid JSON: " + str(error)]
    return validate_surface_spec(document, max_nodes=max_nodes, max_depth=max_depth)

File: frontend/composition/SurfaceValidator.cpp
#include "SurfaceValidator.h"

#include <QJsonArray>
#include <QJsonDocument>
#include <QJsonObject>
#include <QJsonParseError>
#include <QJsonValue>
#include <QMap>
#include <QRegularExpression>
#include <QSet>

#include <cmath>
#include <functional>

namespace {

const QSet<QString> kRootKeys = QSet<QString>()
    << QStringLiteral("version") << QStringLiteral("surfaceId")
    << QStringLiteral("root") << QStringLiteral("nodes");
const QSet<QString> kLayoutKeys = QSet<QString>()
    << QStringLiteral("id") << QStringLiteral("type") << QStringLiteral("children")
    << QStringLiteral("gap") << QStringLiteral("align") << QStringLiteral("justify")
    << QStringLiteral("weight");
const QSet<QString> kBusinessKeys = QSet<QString>()
    << QStringLiteral("id") << QStringLiteral("type") << QStringLiteral("weight");
const QSet<QString> kLayoutTypes = QSet<QString>()
    << QStringLiteral("Row") << QStringLiteral("Column");
const QSet<QString> kGaps = QSet<QString>()
    << QStringLiteral("none") << QStringLiteral("small")
    << QStringLiteral("medium") << QStringLiteral("large");
const QSet<QString> kAligns = QSet<QString>()
    << QStringLiteral("start") << QStringLiteral("center")
    << QStringLiteral("end") << QStringLiteral("stretch");
const QSet<QString> kJustifies = QSet<QString>()
    << QStringLiteral("start") << QStringLiteral("center") << QStringLiteral("end")
    << QStringLiteral("spaceBetween") << QStringLiteral("spaceAround")
    << QStringLiteral("spaceEvenly");

bool hasOnlyKeys(const QJsonObject &object, const QSet<QString> &allowed)
{
    const QStringList keys = object.keys();
    for (QStringList::const_iterator it = keys.constBegin(); it != keys.constEnd(); ++it) {
        if (!allowed.contains(*it)) {
            return false;
        }
    }
    return true;
}

bool hasRequiredKeys(const QJsonObject &object, const QSet<QString> &required)
{
    for (QSet<QString>::const_iterator it = required.constBegin(); it != required.constEnd(); ++it) {
        if (!object.contains(*it)) {
            return false;
        }
    }
    return true;
}

bool isNodeId(const QJsonValue &value)
{
    static const QRegularExpression expression(QStringLiteral("^[A-Za-z][A-Za-z0-9_-]{0,63}$"));
    return value.isString() && expression.match(value.toString()).hasMatch();
}

bool isIntegerInRange(const QJsonValue &value, int minimum, int maximum)
{
    if (!value.isDouble()) {
        return false;
    }
    const double number = value.toDouble();
    return std::floor(number) == number && number >= minimum && number <= maximum;
}

void addError(a2ui::SurfaceValidationResult *result, const QString &message)
{
    result->diagnostics.append(message);
}

} // namespace

namespace a2ui {

SurfaceValidationResult::SurfaceValidationResult()
    : isValid(false)
{
}

SurfaceValidator::Limits::Limits()
    : maximumNodes(32),
      maximumDepth(8)
{
}

SurfaceValidator::SurfaceValidator(const QMap<QString, bool> &catalogMultiplicity,
                                   const Limits &limits)
    : catalogMultiplicity_(catalogMultiplicity),
      limits_(limits)
{
}

QMap<QString, bool> SurfaceValidator::demoCatalog()
{
    QMap<QString, bool> catalog;
    catalog.insert(QStringLiteral("Calculator"), true);
    catalog.insert(QStringLiteral("CalculationHistory"), true);
    catalog.insert(QStringLiteral("CalculationStats"), true);
    catalog.insert(QStringLiteral("Clock"), true);
    catalog.insert(QStringLiteral("NotePad"), true);
    return catalog;
}

SurfaceValidationResult SurfaceValidator::validate(const QByteArray &json) const
{
    QJsonParseError parseError;
    const QJsonDocument document = QJsonDocument::fromJson(json, &parseError);
    if (parseError.error != QJsonParseError::NoError) {
        SurfaceValidationResult result;
        addError(&result, QStringLiteral("invalid JSON: %1").arg(parseError.errorString()));
        return result;
    }
    return validate(document);
}

SurfaceValidationResult SurfaceValidator::validate(const QJsonDocument &document) const
{
    SurfaceValidationResult result;
    if (!document.isObject()) {
        addError(&result, QStringLiteral("document must be an object"));
        return result;
    }

    const QJsonObject rootObject = document.object();
    if (!hasOnlyKeys(rootObject, kRootKeys) || !hasRequiredKeys(rootObject, kRootKeys)) {
        addError(&result, QStringLiteral("document fields must be exactly version, surfaceId, root, nodes"));
    }
    if (!rootObject.value(QStringLiteral("version")).isString()
            || rootObject.value(QStringLiteral("version")).toString() != QStringLiteral("0.1")) {
        addError(&result, QStringLiteral("version must be 0.1"));
    }
    if (!rootObject.value(QStringLiteral("surfaceId")).isString()
            || rootObject.value(QStringLiteral("surfaceId")).toString() != QStringLiteral("main")) {
        addError(&result, QStringLiteral("surfaceId must be main"));
    }
    if (!isNodeId(rootObject.value(QStringLiteral("root")))) {
        addError(&result, QStringLiteral("root must be a valid node ID"));
    }
    if (!rootObject.value(QStringLiteral("nodes")).isArray()) {
        addError(&result, QStringLiteral("nodes must be an array"));
        return result;
    }

    const QString rootId = rootObject.value(QStringLiteral("root")).toString();
    const QJsonArray nodes = rootObject.value(QStringLiteral("nodes")).toArray();
    if (nodes.size() > limits_.maximumNodes) {
        addError(&result, QStringLiteral("node limit exceeded"));
    }

    QMap<QString, QJsonObject> byId;
    QMap<QString, QStringList> childrenById;
    QMap<QString, int> typeCounts;
    for (int index = 0; index < nodes.size(); ++index) {
        const QString label = QStringLiteral("nodes[%1]").arg(index);
        if (!nodes.at(index).isObject()) {
            addError(&result, label + QStringLiteral(" must be an object"));
            continue;
        }
        const QJsonObject node = nodes.at(index).toObject();
        if (!isNodeId(node.value(QStringLiteral("id")))) {
            addError(&result, label + QStringLiteral(" has an invalid id"));
            continue;
        }
        const QString id = node.value(QStringLiteral("id")).toString();
        if (byId.contains(id)) {
            addError(&result, QStringLiteral("duplicate node id: ") + id);
        }
        byId.insert(id, node);

        if (!node.value(QStringLiteral("type")).isString()) {
            addError(&result, label + QStringLiteral(" has an invalid type"));
            childrenById.insert(id, QStringList());
            continue;
        }
        const QString type = node.value(QStringLiteral("type")).toString();
        if (kLayoutTypes.contains(type)) {
            const QSet<QString> required = QSet<QString>()
                << QStringLiteral("id") << QStringLiteral("type") << QStringLiteral("children");
            if (!hasRequiredKeys(node, required)) {
                addError(&result, label + QStringLiteral(" is missing layout fields"));
            }
            if (!hasOnlyKeys(node, kLayoutKeys)) {
                addError(&result, label + QStringLiteral(" contains forbidden layout fields"));
            }
            QStringList childIds;
            const QJsonValue childrenValue = node.value(QStringLiteral("children"));
            if (!childrenValue.isArray()) {
                addError(&result, label + QStringLiteral(".children must be an array of IDs"));
            } else {
                const QJsonArray childArray = childrenValue.toArray();
                if (childArray.size() > limits_.maximumNodes) {
                    addError(&result, label + QStringLiteral(".children exceeds the node limit"));
                }
                for (int childIndex = 0; childIndex < childArray.size(); ++childIndex) {
                    if (!isNodeId(childArray.at(childIndex))) {
                        addError(&result, label + QStringLiteral(".children must contain valid IDs"));
                    } else {
                        childIds.append(childArray.at(childIndex).toString());
                    }
                }
            }
            childrenById.insert(id, childIds);
            if (node.contains(QStringLiteral("gap"))
                    && (!node.value(QStringLiteral("gap")).isString()
                        || !kGaps.contains(node.value(QStringLiteral("gap")).toString()))) {
                addError(&result, label + QStringLiteral(" has an invalid gap"));
            }
            if (node.contains(QStringLiteral("align"))
                    && (!node.value(QStringLiteral("align")).isString()
                        || !kAligns.contains(node.value(QStringLiteral("align")).toString()))) {
                addError(&result, label + QStringLiteral(" has an invalid align"));
            }
            if (node.contains(QStringLiteral("justify"))
                    && (!node.value(QStringLiteral("justify")).isString()
                        || !kJustifies.contains(node.value(QStringLiteral("justify")).toString()))) {
                addError(&result, label + QStringLiteral(" has an invalid justify"));
            }
        } else if (catalogMultiplicity_.contains(type)) {
            const QSet<QString> required = QSet<QString>()
                << QStringLiteral("id") << QStringLiteral("type");
            if (!hasRequiredKeys(node, required)) {
                addError(&result, label + QStringLiteral(" is missing business fields"));
            }
            if (!hasOnlyKeys(node, kBusinessKeys)) {
                addError(&result, label + QStringLiteral(" contains forbidden business fields"));
            }
            typeCounts[type] = typeCounts.value(type, 0) + 1;
            childrenById.insert(id, QStringList());
        } else {
            addError(&result, label + QStringLiteral(" has an unknown or forbidden type"));
            childrenById.insert(id, QStringList());
        }
        if (node.contains(QStringLiteral("weight"))
                && !isIntegerInRange(node.value(QStringLiteral("weight")), 0, 10)) {
            addError(&result, label + QStringLiteral(" has an invalid weight"));
        }
    }

    if (!byId.contains(rootId)) {
        addError(&result, QStringLiteral("root does not reference a declared node"));
    } else if (byId.value(rootId).contains(QStringLiteral("weight"))) {
        addError(&result, QStringLiteral("root must not declare weight"));
    }

    QMap<QString, int> parentCount;
    for (QMap<QString, QJsonObject>::const_iterator it = byId.constBegin(); it != byId.constEnd(); ++it) {
        parentCount.insert(it.key(), 0);
    }
    for (QMap<QString, QStringList>::const_iterator it = childrenById.constBegin();
         it != childrenById.constEnd(); ++it) {
        bool hasPositiveWeight = false;
        const QStringList childIds = it.value();
        for (QStringList::const_iterator child = childIds.constBegin(); child != childIds.constEnd(); ++child) {
            if (!byId.contains(*child)) {
                addError(&result, it.key() + QStringLiteral(" references unknown child ") + *child);
                continue;
            }
            parentCount[*child] = parentCount.value(*child, 0) + 1;
            if (byId.value(*child).value(QStringLiteral("weight")).toInt(0) > 0) {
                hasPositiveWeight = true;
            }
        }
        if (hasPositiveWeight
                && byId.value(it.key()).value(QStringLiteral("justify")).toString(QStringLiteral("start"))
                    != QStringLiteral("start")) {
            addError(&result, it.key() + QStringLiteral(" combines positive child weight with non-start justify"));
        }
    }
    for (QMap<QString, int>::const_iterator it = parentCount.constBegin(); it != parentCount.constEnd(); ++it) {
        const int expected = it.key() == rootId ? 0 : 1;
        if (it.value() != expected) {
            addError(&result, QStringLiteral("%1 has %2 parents; expected %3")
                     .arg(it.key()).arg(it.value()).arg(expected));
        }
    }

    QSet<QString> visited;
    QSet<QString> active;
    std::function<void(const QString &, int)> visit = [&](const QString &id, int depth) {
        if (active.contains(id)) {
            addError(&result, QStringLiteral("cycle detected at ") + id);
            return;
        }
        if (visited.contains(id) || !byId.contains(id)) {
            return;
        }
        if (depth > limits_.maximumDepth) {
            addError(&result, QStringLiteral("depth limit exceeded at ") + id);
        }
        active.insert(id);
        const QStringList childIds = childrenById.value(id);
        for (QStringList::const_iterator it = childIds.constBegin(); it != childIds.constEnd(); ++it) {
            visit(*it, depth + 1);
        }
        active.remove(id);
        visited.insert(id);
    };
    if (byId.contains(rootId)) {
        visit(rootId, 1);
    }
    QSet<QString> unreachable;
    for (QMap<QString, QJsonObject>::const_iterator it = byId.constBegin(); it != byId.constEnd(); ++it) {
        if (!visited.contains(it.key())) {
            unreachable.insert(it.key());
        }
    }
    if (!unreachable.isEmpty()) {
        QStringList ids = unreachable.values();
        ids.sort();
        addError(&result, QStringLiteral("unreachable nodes: ") + ids.join(QStringLiteral(", ")));
    }

    for (QMap<QString, bool>::const_iterator it = catalogMultiplicity_.constBegin();
         it != catalogMultiplicity_.constEnd(); ++it) {
        if (!it.value() && typeCounts.value(it.key(), 0) > 1) {
            addError(&result, it.key() + QStringLiteral(" does not allow multiple instances"));
        }
    }

    result.isValid = result.diagnostics.isEmpty();
    return result;
}

} // namespace a2ui
